activestorage/lib/active_storage/attached/changes/create_one.rb

```diff
@@ -69,7 +69,21 @@ def find_or_build_blob
             ).symbolize_keys
           )
         when String
-          ActiveStorage::Blob.find_signed!(attachable, record: record)
```

Comment:
I'm hesitant to believe that this code path will only ever be used for direct uploads. Users could try attaching by signed ID via database seeding for example. There may also be legitimate use-cases for using one blob for multiple attachment relations.

How can we conditionalize this further to only check when the attachable is meant to be a direct upload? The fact that we had to turn on the new feature for all of our one + many attachment tests is an indicator that we're breaking contract on other code paths.

Reply:
Hey @gmcgibbon, we gave some thought to your comments and we think that breaking the API is inevitable if we are to fix the security vulnerability but we should also contemplate older apps running on production right now. To consider these older apps we agree this should be a configuration to allow them to keep their current behavior and since, as you mentioned, the path is more widely used we thought about renaming the purpose from `:direct_upload` to `:private_id`.

The configuration will be set to use `:private_id` by default on new Rails 7+ apps prioritizing security and `:blob_id` on older ones. To ensure a smooth migration we also included a `:private_id_with_fallback` value which will sign direct uploads with `:private_id` but accept both `:blob_id` and `:private_id` until they change it to the more secure mode `:private_id`. We've included these changes along with an explanation in the CHANGELOG and migration guides as per @zzak's comments :)